### PR TITLE
feature: S3C-2946 get and head object support object lock

### PR DESCRIPTION
**Pull request by ilkescality**
Extends GET and HEAD Object APIs to return object lock related information for the object if any set on the object's metadata.

When an object version have object lock related information stored in its metadata, GET Object and HEAD Object APIs should return these in the response; sample GET/HEAD Object response for a locked object:
```
{   
    "AcceptRanges": "bytes",
    "LastModified": "2020-06-03T03:26:39+00:00",
    "ContentLength": 0,
    "ETag": "\"d41d8cd98f00b204e9800998ecf8427e\"",
    "VersionId": "39383430383834353230303133383939393939395247303031202030",
    "Metadata": {},
    "ObjectLockMode": "GOVERNANCE",
    "ObjectLockRetainUntilDate": "2021-12-12T00:00:34+00:00",
    "ObjectLockLegalHoldStatus": "ON"
}
```

### CODE REVIEW

File: lib/utilities/collectResponseHeaders.js [hunk 1]
```diff
@@ -75,6 +75,17 @@ function collectResponseHeaders(objectMD, corsHeaders, versioningCfg,
         responseMetaHeaders['x-amz-tagging-count'] =
           Object.keys(objectMD.tags).length;
     }
+    if (objectMD.retentionInfo && (objectMD.retentionInfo.retainUntilDate
```
Comment (miniscruff) [inline lib/utilities/collectResponseHeaders.js]: the parenthesis around conditions is not required here as left to right ANDS are the same.

File: lib/utilities/collectResponseHeaders.js [hunk 2]
```diff
@@ -75,6 +75,17 @@ function collectResponseHeaders(objectMD, corsHeaders, versioningCfg,
...
+        && objectMD.retentionInfo.mode)) {
+            responseMetaHeaders['x-amz-object-lock-retain-until-date']
```
Comment (miniscruff) [inline lib/utilities/collectResponseHeaders.js]: Having these on two lines is hard to read, I am sure eslint will complain at you if you put it on one line but like, is there something we can do here?

File: tests/unit/api/objectGet.js [hunk 3]
```diff
@@ -75,6 +75,155 @@ describe('objectGet API', () => {
...
+    it('should get the object metadata with valid retention info', done => {
+        bucketPut(authInfo, testPutBucketRequestObjectLock, log, () => {
+            const request = testPutObjectReqRetention(testDate, 'GOVERNANCE');
+            objectPut(authInfo, request, undefined,
+                log, (err, headers) => {
+                    assert.ifError(err);
+                    assert.strictEqual(headers.ETag, `"${correctMD5}"`);
+                    objectGet(authInfo, testGetRequest, false, log,
+                        (err, res, headers) => {
+                            assert.ifError(err);
```
Comment (miniscruff) [inline tests/unit/api/objectGet.js]: can anything be done about the amount of whitespace on these lines? The best I could tell was merging lines 106/107 and 110/111 but even then its quite a lot of empty space. Makes it a little hard to read.

File: tests/unit/api/objectGet.js [hunk 4]
```diff
@@ -75,6 +75,155 @@ describe('objectGet API', () => {
...
+        bucketName,
+        namespace,
+        headers: {
+            host: `${bucketName}.s3.amazonaws.com`,
+            'x-amz-object-lock-enabled': true,
+        },
+        url: `/${bucketName}`,
+    };
+
+    const testPutObjectReqRetention = (date, mode) => new DummyRequest({
```
Comment (miniscruff) [inline tests/unit/api/objectGet.js]: Maybe a name that dictates it just creates a dummy request like, `createPutObjectDummyRequest` or `createPutDummy` or `createPutRequest`. I thought this was an actual test the first time I read it.

File: tests/unit/api/objectGet.js [hunk 5]
```diff
@@ -75,6 +75,155 @@ describe('objectGet API', () => {
...
+                                'GOVERNANCE');
+                            assert.strictEqual(headers.ETag,
+                                `"${correctMD5}"`);
+                            done();
+                        });
+                });
+        });
+    });
+
+    const testPutObjectReqLegalHold = legalHold => new DummyRequest({
```
Comment (miniscruff) [inline tests/unit/api/objectGet.js]: Same as above.

**Review by miniscruff — APPROVED** [review]
Looks fine but small comments might improve it "might".

File: lib/utilities/collectResponseHeaders.js (hunk 1 above)
Comment (ilkescality) [reply lib/utilities/collectResponseHeaders.js]: At first I did without those, then I confused myself :) gonna remove them back.

File: lib/utilities/collectResponseHeaders.js (hunk 2 above)
Comment (ilkescality) [reply lib/utilities/collectResponseHeaders.js]: Yeah I agree about this, eslint will definitely complain but I'll see if there's anything that can be done to make this look better - either bypassing lint or I will think of something else.

File: tests/unit/api/objectGet.js (hunk 3 above)
Comment (ilkescality) [reply tests/unit/api/objectGet.js]: I'll work on something about this, bothered me while writing it as well tbh.

File: lib/utilities/collectResponseHeaders.js (hunk 2 above)
Comment (rahulreddy) [reply lib/utilities/collectResponseHeaders.js]: What I did when I ran into this was to move the conditional to a variable, for example,
```js
const objectLocked = objectMD.retentionInfo && objectMD.retentionInfo.retainUntilDate
        && objectMD.retentionInfo.mode;
if (objectLocked) {
  // do the logic
}
```

Comment (miniscruff) [reply lib/utilities/collectResponseHeaders.js]: I am a huge fan that bools should always represent the "positive" form. It overall simplifies the logic and terms. The positive form is not always the default but generally is. For example, even though all this logic has to do with the objects being locked. The bool that dictates if the object is in fact locked might want to be false if it is. Something like `objectOpen` or `objectEditable` ( kinda guessing since I do not know any more terms here that might be more standard ). Note that "unlocked" is not a good name as its effectively a double negative. 

Consider a variable that dictates if a user can go through a portal ( some game idk ).
```js
if (!portalClosed) {
  // go in
}
// OR
if (portalOpen) {
  // go in
}
```

The first says we can go in the portal if it is not closed. Or the second is just it is open. The positive form here is using the portal for portal things.

This is a little difficult at times for conditions that are sort of "blocker" conditions. Just as in this case the purpose of a lock is to block something from happening. If a door had a lock and a deadbolt we might consider the lock "engaged" to be the lock would block the door from opening. So the door could be opened if all locks are not engaged. However, since the primary function of an actual lock is to prevent things from happening its positive value is true if the lock is engaged.

```javascript
const canOpenDoor = !deadbolt.engaged && !lock.engaged;
if (canOpenDoor) {
  // do good things
}
// or
const doorLocked = deadbolt.engaged || lock.engaged;
if (!doorLocked) {
  // double negative
}
```

Also prefixing booleans with is, can, has or similar word is also very powerful as it makes conditions turn almost into sentences. Compare "if I can open the door then open the door" to "if not door locked open door". 

tl;dr use positive forms for booleans, rename locked to something like open or writable or something in the positive form. 

This might be a lot of changes since so much work was put in to add "locking" but just like mentioning this sort of stuff. I may also have some terms wrong and the lock here is actually the positive case for something in which case carry on...

**Review by anurag4DSB — APPROVED** [review]
I like the intensional good namings of variables and methods :)

File: lib/utilities/collectResponseHeaders.js (hunk 2 above)
Comment (anurag4DSB) [reply lib/utilities/collectResponseHeaders.js]: +1

Comment (rahulreddy) [reply lib/utilities/collectResponseHeaders.js]: I agree on the usage of booleans to be in positive form, it avoids double negations and what not. I was just trying to illustrate that you can extract the evaluation logic into a var and use that for conditionals. I will let Ilke chose the naming, I used `lock` as we are working with object locking mechanism.

Comment (miniscruff) [reply lib/utilities/collectResponseHeaders.js]: oh yea my response wasn't an alternative just an additional comment on booleans in general.

File: tests/unit/api/objectGet.js (hunk 5 above)
Comment (miniscruff) [reply tests/unit/api/objectGet.js]: @ilkescality this was not addressed?

File: lib/utilities/collectResponseHeaders.js (hunk 2 above)
Comment (ilkescality) [reply lib/utilities/collectResponseHeaders.js]: @miniscruff @rahulreddy @anurag4DSB thanks for the reviews
@miniscruff thanks for the thorough review; great suggestions. I realized I forgot to reply here after making the changes yesterday. I made the change to make positive use of booleans and named as `hasRetentionInfo` (I thought of doing it like this initially but then ended up keeping it similar with other if statements in the file, definitely looks a lot more readable now though).

In terms of renaming `locked` to something like open/editable / writable / something in the positive form; I think it’s a good practice in general too but for the object lock context, as we inevitably use the term `lock / locked` in bunch of places, (such as error messages, headers, API method names etc.) I’m guessing introducing another term to be a synonym for ‘lock’ might be confusing overall (in terms of the object lock feature at a holistic level) so I think it might probably worth to keep it.

File: tests/unit/api/objectGet.js (hunk 5 above)
Comment (ilkescality) [reply tests/unit/api/objectGet.js]: Oh yeah seems like I missed that one, will change this too.

File: lib/utilities/collectResponseHeaders.js (hunk 2 above)
Comment (miniscruff) [reply lib/utilities/collectResponseHeaders.js]: That's fair, I couldn't really come up with anything better than locked in my examples. Just know that at some points when working with locking you may need to swap the bool to make sense. But that is fine, its more a general idea and not a rule.

**Review by dora-korpar — APPROVED** [review]

**Comment by ilkescality** [comment]
@bert-e approve

**Comment by ilkescality** [comment]
@bert-e reset

**Comment by ilkescality** [comment]
@bert-e approve
